### ISSUE TITLE
[Phase 1] Data Foundation & Preprocessing

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/multi-label-doc-classifier#2]
## Summary

Build the foundational data loading and text preprocessing pipeline. This phase creates the infrastructure for all subsequent phases.

**Parent Issue**: #1

## Objectives

- Load documents from various file formats (JSON, CSV)
- Clean and normalize text documents
- Build TF-IDF feature matrices with sublinear scaling
- Parse and validate hierarchical label taxonomies

## System Context

```
src/
├── features/
│   ├── preprocessor.py    # NEW - Text cleaning and tokenization
│   └── vectorizer.py      # NEW - TF-IDF vectorization wrapper
└── utils/
    └── data_loader.py     # NEW - Document and label loading

data/
└── labels/
    └── taxonomy.json      # NEW - Label hierarchy schema
```

### Data Loading
- [ ] Create `src/utils/data_loader.py`
- [ ] Implement `load_documents(path: str) -> List[Document]`
- [ ] Support JSON format: `[{"text": "...", "labels": [...]}]`
- [ ] Support CSV format with text and label columns
- [ ] Handle encoding issues (UTF-8, Latin-1)

### Text Preprocessing
- [ ] Create `src/features/preprocessor.py`
- [ ] Implement `TextPreprocessor` class with:
  - `lowercase: bool = True`
  - `remove_punctuation: bool = True`
  - `remove_numbers: bool = False`
  - `min_token_length: int = 2`
- [ ] Add `process(text: str) -> str` method
- [ ] Add `process_batch(texts: List[str]) -> List[str]` method

### TF-IDF Vectorization
- [ ] Create `src/features/vectorizer.py`
- [ ] Implement `DocumentVectorizer` wrapping sklearn's TfidfVectorizer
- [ ] Configure sublinear_tf=True for sublinear scaling
- [ ] Support custom n-gram ranges for domain terminology
- [ ] Return sparse matrices for memory efficiency

### Label Taxonomy
- [ ] Create `data/labels/taxonomy.json` schema
- [ ] Implement taxonomy validation in `src/utils/data_loader.py`
- [ ] Support hierarchical parent-child relationships

### TextPreprocessor Skeleton

```python
# src/features/preprocessor.py
import re
import string
from typing import List

class TextPreprocessor:
    def __init__(
        self,
        lowercase: bool = True,
        remove_punctuation: bool = True,
        remove_numbers: bool = False,
        min_token_length: int = 2
    ):
        self.lowercase = lowercase
        self.remove_punctuation = remove_punctuation
        self.remove_numbers = remove_numbers
        self.min_token_length = min_token_length

        # Precompile patterns
        self._punctuation_pattern = re.compile(f'[{re.escape(string.punctuation)}]')
        self._number_pattern = re.compile(r'\d+')
        self._whitespace_pattern = re.compile(r'\s+')

    def process(self, text: str) -> str:
        if not text:
            return ""

        result = text
        if self.lowercase:
            result = result.lower()
        if self.remove_punctuation:
            result = self._punctuation_pattern.sub(' ', result)
        if self.remove_numbers:
            result = self._number_pattern.sub(' ', result)

        result = self._whitespace_pattern.sub(' ', result).strip()

        if self.min_token_length > 1:
            tokens = [t for t in result.split() if len(t) >= self.min_token_length]
            result = ' '.join(tokens)

        return result

    def process_batch(self, texts: List[str]) -> List[str]:
        return [self.process(text) for text in texts]
```

### DocumentVectorizer Skeleton

```python
# src/features/vectorizer.py
from sklearn.feature_extraction.text import TfidfVectorizer
from scipy.sparse import csr_matrix
from typing import List, Optional

class DocumentVectorizer:
    def __init__(
        self,
        max_features: int = 5000,
        ngram_range: tuple = (1, 2),
        min_df: int = 2,
        max_df: float = 0.95
    ):
        self.vectorizer = TfidfVectorizer(
            max_features=max_features,
            ngram_range=ngram_range,
            min_df=min_df,
            max_df=max_df,
            sublinear_tf=True,  # Use 1 + log(tf) instead of tf
            dtype='float32'     # Save memory
        )

    def fit_transform(self, documents: List[str]) -> csr_matrix:
        return self.vectorizer.fit_transform(documents)

    def transform(self, documents: List[str]) -> csr_matrix:
        return self.vectorizer.transform(documents)

    def get_feature_names(self) -> List[str]:
        return self.vectorizer.get_feature_names_out().tolist()
```

## Verification

```python
# Test loading and vectorizing
from src.utils.data_loader import load_documents
from src.features.preprocessor import TextPreprocessor
from src.features.vectorizer import DocumentVectorizer

docs = load_documents("data/raw/sample.json")
preprocessor = TextPreprocessor()
clean_docs = preprocessor.process_batch([d.text for d in docs])
vectorizer = DocumentVectorizer(max_features=1000)
X = vectorizer.fit_transform(clean_docs)

assert X.shape[0] == len(docs), "Row count should match document count"
assert X.shape[1] <= 1000, "Feature count should respect max_features"
print(f"Vectorized {X.shape[0]} documents into {X.shape[1]} features")
```

## Definition of Done

- [ ] All unit tests pass (`pytest tests/test_preprocessor.py tests/test_vectorizer.py -v`)
- [ ] Can load 1000+ documents without memory errors
- [ ] Vectorizer produces consistent output on same input (deterministic)
- [ ] Label taxonomy JSON schema is documented
- [ ] Code follows project style (black, flake8 clean)

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/features/preprocessor.py` | - | Create | Text preprocessing class |
| `src/features/vectorizer.py` | - | Create | TF-IDF vectorization wrapper |
| `src/utils/data_loader.py` | - | Create | Document loading functions |
| `data/labels/taxonomy.json` | - | Create | Label hierarchy schema |
| `tests/test_preprocessor.py` | - | Create | Preprocessor unit tests |
| `tests/test_vectorizer.py` | - | Create | Vectorizer unit tests |
